terraform-plans/templates/github/charm_check.yaml.tftpl

```diff
@@ -72,21 +72,53 @@ jobs:
     - name: Run unit tests
       run: tox -e unit
 
-  func:
+  build:
     needs:
       - lint
       - unit
+    name: build charms
+    runs-on: $${{ matrix.runs-on }}
+    timeout-minutes: 120
+    strategy:
+      fail-fast: false
+      matrix:
+        runs-on: ${ runs_on }
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+
+      - uses: canonical/craft-actions/charmcraft/pack@main
+        id: build
+        with:
+          verbosity: verbose
+          channel: "${ charmcraft_channel }"
```

Comment:
Can we check the details of how this caches things here? I have some doubts:

- if there are multiple jobs, they will be likely running in parallel, so they won't be able to benefit from the cache of the other
- we're only going to be running multiple jobs if we need to build on different architectures. I'm worried that cache artefacts from one architecture will be different to that of another architecture.

Maybe the cache is a good idea; I haven't looked far into the details. But my personal thought on this is I'd rather ensure the charm is built from a true clean reproducible environment, because this will be the production charm we're publishing. If the build takes a bit longer, it's still worth it imo.

Reply:
it's based on the following https://github.com/canonical/craft-actions/blob/main/charmcraft/pack/action.yaml#L58

Reply:
I did not think about the multi-arch scenario. I was thinking about e.g. rebuilding charm during a PR, where you can constantly pushing different commit without changing the dependency

Reply:
Ah of course, I wasn't sure how long the caches lived - that could definitely be helpful if it works across test runs. Maybe we need to look into setting a custom cache key if we're going to use the cache? To ensure that it depends on the architecture (if that applies) and any requirements files, etc. we use that might not be picked up by the default cache key.

Although even then, if we want to be extra careful, dependencies can change even if our requirements.txt doesn't change, because we're not using lock files. It feels a little risky to me if we're using this for the production charm uploads. :sweat_smile:

Reply:
Or maybe something like only using the cache for code running in PRs, or PRs in draft mode when there may be commits pushed frequently? That might help for saving time there, but ensuring a clean environment for the production build?

Reply:
Yeah, thanks for that, maybe we should just rollback we previously did for now. There's still at least on unclear points from me based on your question: does this work properly in multi-arch case

Reply:
btw charmcraft team recommend using this https://canonical-charmcraft.readthedocs-hosted.com/en/stable/howto/build-guides/shared-cache/#on-github